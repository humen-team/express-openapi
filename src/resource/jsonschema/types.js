/* Cast a JSON value to a JSON Schema primitive type.
 */
export function castInputValue(value, type) {
    if (type === 'integer' || type === 'long') {
        return parseInt(value, 10);
    }
    if (type === 'float' || type === 'double') {
        return parseFloat(value);
    }
    if (type === 'boolean') {
        return value.toLowerCase() === 'true' || value === '1';
    }
    if (type === 'date' || type === 'dateTime') {
        return new Date(value);
    }
    if (type === 'byte') {
        return Buffer.from(value, 'base64');
    }
    if (type === 'binary') {
        return Buffer.from(value, 'binary');
    }
    return value;
}

/* Cast a value to JSON.
 */
export function castOutputValue(value) {
    if (value instanceof Date) {
        return value.toISOString();
    }
    return value;
}
