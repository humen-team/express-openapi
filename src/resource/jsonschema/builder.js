import { isPlainObject, mapValues, merge } from 'lodash';

import { OPENAPI_2_0 } from '../../constants';

/* Does this schema have a nullable type?
 */
export function isNullable(object) {
    const { type } = object;
    return type === 'null' || (type?.includes && type.includes(null));
}

/* Build valid OpenAPI from JSON Schema.
 */
export default function buildOpenAPI(object, openapiVersion, path = '') {
    // recurse into lists
    if (Array.isArray(object)) {
        return object.map(
            (item) => buildOpenAPI(item, openapiVersion, `${path}[]`),
        );
    }

    const nullable = openapiVersion === OPENAPI_2_0 ? 'x-nullable' : 'nullable';

    // recurse into objects
    if (isPlainObject(object)) {
        return merge(
            // special case #1: nullability can be encoded
            object.type ? {
                [nullable]: isNullable(object),
            } : {},
            mapValues(
                object,
                (value, key) => {
                    // special case #2: types cannot be lists or contain `null`
                    if (key === 'type' && !isPlainObject(value)) {
                        return value?.filter
                            ? value.filter((item) => item !== null)[0]
                            : value;
                    }
                    // special case #3: items is a reference to another part of the document
                    if (key === '$ref') {
                        return value.build(openapiVersion);
                    }
                    return buildOpenAPI(value, openapiVersion, `${path}.${key}`);
                },
            ),
        );
    }

    // return primitives
    return object;
}
