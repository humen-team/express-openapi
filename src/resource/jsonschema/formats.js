import { validate } from 'uuid';

export function uuidFormat(input) {
    return input === null || validate(input);
}

export function userIdFormat(input) {
    return input === 'me' || uuidFormat(input);
}
