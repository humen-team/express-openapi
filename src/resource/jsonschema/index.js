import JSONSchemaResource from './resource';

export default JSONSchemaResource;
