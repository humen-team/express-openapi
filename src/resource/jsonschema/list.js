import JSONSchemaResource from './resource';

/* A resource that represents a list of another resource.
 */
export default class JSONSchemaListResource extends JSONSchemaResource {
    constructor(item) {
        const schema = JSONSchemaListResource.createSchema(item);
        super(schema, item.options);
        this.item = item;
    }

    /* Create a new JSONSchema represention for a list of items.
     */
    static createSchema(item) {
        const { config } = item;
        const { namingStrategy, pagingStrategy } = config;
        const schema = {
            id: namingStrategy.toListResourceId(item.id),
            type: 'object',
            properties: {
                items: {
                    type: 'array',
                    items: {
                        $ref: item.toRef(),
                    },
                },
            },
            required: [
                'items',
            ],
        };

        // enable response to report on current offset/limit paging
        if (pagingStrategy.hasOffsetLimit) {
            schema.properties.limit = {
                type: 'integer',
            };
            schema.properties.offset = {
                type: 'integer',
            };
        }
        // enable response to report on total collection count
        if (pagingStrategy.hasCount) {
            schema.properties.count = {
                type: 'integer',
            };
            schema.required.push('count');
        }

        return schema;
    }
}
