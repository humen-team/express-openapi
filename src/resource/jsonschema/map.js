import { get, mapValues } from 'lodash';

/* Map a function over an object and a schema.
 */
export default function mapSchema(value, schema, registry, func, path = '') {
    if (!schema) {
        return func(value, null);
    }

    const { $ref, type } = schema;

    if ($ref) {
        // XXX resolved Reference
        return mapSchema(
            value,
            registry[$ref],
            registry,
            func,
            `${path}.$ref`,
        );
    }

    if (type === 'array') {
        return value.map(
            (child) => mapSchema(
                child,
                get(schema, 'items'),
                registry,
                func,
                `${path}.items`,
            ),
        );
    }

    if (type === 'object') {
        return mapValues(
            value,
            (child, key) => mapSchema(
                child,
                get(schema, `properties.${key}`),
                registry,
                func,
                `${path}.properties.${key}`,
            ),
        );
    }

    return func(value, schema);
}
