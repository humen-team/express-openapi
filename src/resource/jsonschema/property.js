import { merge, omit } from 'lodash';

import Reference from '../reference';
import { OPENAPI_2_0 } from '../../constants';

/* Does this schema have a nullable type?
 */
export function isNullable(schema) {
    const { type } = schema;
    return type === 'null' || (type?.includes && type.includes(null));
}

/* Build a schema property as valid OpenAPI.
 */
export function buildProperty(schema, openapiVersion) {
    const { items, type } = schema;
    const nullable = openapiVersion === OPENAPI_2_0 ? 'x-nullable' : 'nullable';

    // special case #1: nullability is encoded differently
    const converted = {
        [nullable]: isNullable(schema),
    };

    // special case #2: types cannot be lists or contain `null`
    converted.type = type?.filter
        ? type.filter((item) => item !== null)[0]
        : type;

    // special case #3: items is a reference to another part of the document
    if (items instanceof Reference) {
        converted.items = items.build(openapiVersion);
    } else {
        converted.items = items;
    }

    return merge(
        omit(schema, ['items', 'type']),
        converted,
    );
}
