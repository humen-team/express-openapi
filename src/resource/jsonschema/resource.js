import {
    difference,
    intersection,
    isUndefined,
    mapValues,
    merge,
    omit,
    pick,
    union,
} from 'lodash';

import UnprocessableEntity from '../../errors/unprocessable_entity';
import Resource from '../resource';
import mapSchema from './map';
import { buildProperty } from './property';
import { castInputValue, castOutputValue } from './types';
import createValidator from './validator';

/* Defines an OpenAPI compatible resource using JSONSchema.
 */
export default class JSONSchemaResource extends Resource {
    constructor(schema, options) {
        super(options);
        // default to stricter validation
        const { additionalProperties, type } = schema;
        this.schema = {
            additionalProperties: additionalProperties === undefined
                ? false
                : additionalProperties,
            type: type || 'object',
            ...schema,
        };
        this.validator = null;
    }

    get id() {
        return this.config.namingStrategy.toName(this.schema.id);
    }

    get properties() {
        return this.schema.properties;
    }

    get required() {
        return this.schema.required;
    }

    get type() {
        return this.schema.type;
    }

    get keys() {
        return Object.keys(this.schema.properties || {});
    }

    /* Convert this resource to an OpenAPI compatible definition.
     *
     * While OpenAPI leverages JSONSchema, it isn't 100% compatible; some conversion is required.
     */
    build(openapiVersion) {
        const { properties, required, type } = this.schema;
        return {
            // omit the `id` field
            properties: mapValues(
                properties,
                (value) => buildProperty(value, openapiVersion),
            ),
            required,
            type,
        };
    }

    /* Convert JSON data to this resource's expected types.
     */
    castInput(input) {
        return mapSchema(
            input,
            this.schema,
            // NB: keep unmapped values as additionalProperties
            (object, schema) => (schema ? castInputValue(object, schema) : object),
        );
    }

    /* Convert this resource's data to JSON types.
     */
    castOutput(output) {
        return mapSchema(
            output,
            this.schema,
            // NB: omit unmapped values to avoid additionalProperties
            (object, schema) => (schema ? castOutputValue(object, schema) : undefined),
        );
    }

    toList() {
        // avoid circular dependency by deferring import
        const JSONSchemaResourceList = require('./list').default; // eslint-disable-line global-require
        return new JSONSchemaResourceList(this);
    }

    /* Validate data against this JSONSchema schema.
     */
    validate(data) {
        if (!this.validator) {
            this.validator = createValidator(this);
        }

        const result = this.validator.validate(data);

        if (result.errors.length) {
            // NB: we could report on mutiple errors
            const { property, message } = result.errors[0];
            throw new UnprocessableEntity({
                message: `${property} ${message}`,
            });
        }
        return data;
    }

    /* Create a resource, requiring all properties.
     */
    static all(options = {}) {
        const { id = this.id, properties = {} } = options;
        return new JSONSchemaResource({
            id,
            properties,
            required: Object.keys(properties),
            type: 'object',
        });
    }

    /* Add some properties to a resource.
     */
    merge(options = {}) {
        const {
            id = this.id,
            properties = {},
            required = [],
        } = options;
        return new JSONSchemaResource({
            id,
            properties: merge(properties, this.properties),
            required: union(required, this.required),
            type: 'object',
        });
    }

    /* Remove some properties from a resource.
     */
    omit(options = {}) {
        const {
            id = this.id,
            properties = [],
            required,
        } = options;
        return new JSONSchemaResource({
            id,
            properties: omit(this.properties, properties),
            required: required || difference(this.required, properties),
            type: 'object',
        });
    }

    /* Pick some properties from a resource.
     */
    pick(options = {}) {
        const {
            id = this.id,
            properties = [],
            required,
        } = options;
        return new JSONSchemaResource({
            id,
            properties: pick(this.properties, properties),
            required: isUndefined(required) ? intersection(this.required, properties) : required,
            type: 'object',
        });
    }
}
