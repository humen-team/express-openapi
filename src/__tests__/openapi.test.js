import { readFileSync } from 'fs';
import { Validator } from 'jsonschema';
import request from 'supertest';

import createApp from './app';

describe('app', () => {
    const openapi2 = JSON.parse(readFileSync('schemas/openapi/2.0/schema.json'));
    const openapi3 = JSON.parse(readFileSync('schemas/openapi/3.0/schema.json'));
    const jsonSchema = JSON.parse(readFileSync('schemas/json-schema/draft-04/schema.json'));
    const validator = new Validator();
    validator.addSchema(jsonSchema);

    describe('OpenAPI', () => {
        describe('2.0', () => {
            it('returns 200', async () => {
                const app = createApp();
                const response = await request(app).get('/openapi/2.0');
                expect(response.statusCode).toEqual(200);
                expect(response.body).toMatchSnapshot();
                validator.validate(response.body, openapi2, { throwError: true });
            });
        });
        describe('3.0.0', () => {
            it('returns 200', async () => {
                const app = createApp();
                const response = await request(app).get('/openapi/3.0.0');
                expect(response.statusCode).toEqual(200);
                expect(response.body).toMatchSnapshot();
                validator.validate(response.body, openapi3, { throwError: true });
            });
        });
    });
});
