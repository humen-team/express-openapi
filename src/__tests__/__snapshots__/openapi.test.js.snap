// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app OpenAPI 2.0 returns 200 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "CommandInput": Object {
      "properties": Object {
        "request": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "request",
      ],
      "type": "object",
    },
    "CommandOutput": Object {
      "properties": Object {
        "response": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "response",
      ],
      "type": "object",
    },
    "CountFoo": Object {
      "properties": Object {},
      "type": "object",
    },
    "CreateFoo": Object {
      "properties": Object {
        "bar": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "bar",
      ],
      "type": "object",
    },
    "Error": Object {
      "properties": Object {
        "message": Object {
          "type": "string",
          "x-nullable": false,
        },
        "status": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
    },
    "Foo": Object {
      "properties": Object {
        "bar": Object {
          "type": "string",
          "x-nullable": false,
        },
        "id": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "bar",
        "id",
      ],
      "type": "object",
    },
    "FooFile": Object {
      "properties": Object {
        "id": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "id",
      ],
      "type": "object",
    },
    "FooList": Object {
      "properties": Object {
        "count": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "items": Object {
          "items": Object {
            "$ref": "#/definitions/Foo",
          },
          "type": "array",
          "x-nullable": false,
        },
        "limit": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "offset": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
      "required": Array [
        "items",
        "count",
      ],
      "type": "object",
    },
    "QueryOutput": Object {
      "properties": Object {
        "response": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "response",
      ],
      "type": "object",
    },
    "ReplaceFoo": Object {
      "properties": Object {
        "bar": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "bar",
      ],
      "type": "object",
    },
    "UpdateFoo": Object {
      "properties": Object {
        "bar": Object {
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "bar",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "paths": Object {
    "/command": Object {
      "post": Object {
        "consumes": Array [
          "application/json",
        ],
        "operationId": "mycommand",
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/CommandInput",
            },
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "CommandOutput as JSON",
            "schema": Object {
              "$ref": "#/definitions/CommandOutput",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
      },
    },
    "/foo": Object {
      "get": Object {
        "operationId": "search",
        "parameters": Array [
          Object {
            "in": "query",
            "name": "offset",
            "type": "integer",
          },
          Object {
            "in": "query",
            "name": "limit",
            "type": "integer",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "FooList as JSON",
            "schema": Object {
              "$ref": "#/definitions/FooList",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "head": Object {
        "operationId": "count",
        "parameters": Array [],
        "responses": Object {
          "204": Object {
            "description": "Empty response",
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "post": Object {
        "consumes": Array [
          "application/json",
        ],
        "operationId": "create",
        "parameters": Array [
          Object {
            "in": "body",
            "name": "foo",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/CreateFoo",
            },
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "201": Object {
            "description": "Foo as JSON",
            "schema": Object {
              "$ref": "#/definitions/Foo",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
    },
    "/foo/{fooId}": Object {
      "delete": Object {
        "operationId": "delete",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "type": "string",
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "Empty response",
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "get": Object {
        "operationId": "retrieve",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "Foo as JSON",
            "schema": Object {
              "$ref": "#/definitions/Foo",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "patch": Object {
        "consumes": Array [
          "application/json",
        ],
        "operationId": "update",
        "parameters": Array [
          Object {
            "in": "body",
            "name": "foo",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/UpdateFoo",
            },
          },
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "Foo as JSON",
            "schema": Object {
              "$ref": "#/definitions/Foo",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "put": Object {
        "consumes": Array [
          "application/json",
        ],
        "operationId": "replace",
        "parameters": Array [
          Object {
            "in": "body",
            "name": "foo",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/ReplaceFoo",
            },
          },
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "Foo as JSON",
            "schema": Object {
              "$ref": "#/definitions/Foo",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "foo",
        ],
      },
    },
    "/foo_file": Object {
      "post": Object {
        "operationId": "createfile",
        "parameters": Array [],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "201": Object {
            "description": "FooFile as JSON",
            "schema": Object {
              "$ref": "#/definitions/FooFile",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "fooFile",
        ],
      },
    },
    "/foo_file/{fooFileId}": Object {
      "get": Object {
        "operationId": "retrievefile",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooFileId",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "image/png",
        ],
        "responses": Object {
          "200": Object {
            "description": "fooFile file content",
            "schema": Object {
              "type": "file",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "fooFile",
        ],
      },
      "put": Object {
        "operationId": "replacefile",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooFileId",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "FooFile as JSON",
            "schema": Object {
              "$ref": "#/definitions/FooFile",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "fooFile",
        ],
      },
    },
    "/query": Object {
      "get": Object {
        "operationId": "myquery",
        "parameters": Array [
          Object {
            "in": "query",
            "name": "request",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "QueryOutput as JSON",
            "schema": Object {
              "$ref": "#/definitions/QueryOutput",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
      },
    },
  },
  "schemes": Array [
    "http",
  ],
  "swagger": "2.0",
}
`;

exports[`app OpenAPI 3.0.0 returns 200 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "CommandInput": Object {
        "properties": Object {
          "request": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "request",
        ],
        "type": "object",
      },
      "CommandOutput": Object {
        "properties": Object {
          "response": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "response",
        ],
        "type": "object",
      },
      "CountFoo": Object {
        "properties": Object {},
        "type": "object",
      },
      "CreateFoo": Object {
        "properties": Object {
          "bar": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "bar",
        ],
        "type": "object",
      },
      "Error": Object {
        "properties": Object {
          "message": Object {
            "nullable": false,
            "type": "string",
          },
          "status": Object {
            "nullable": false,
            "type": "integer",
          },
        },
      },
      "Foo": Object {
        "properties": Object {
          "bar": Object {
            "nullable": false,
            "type": "string",
          },
          "id": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "bar",
          "id",
        ],
        "type": "object",
      },
      "FooFile": Object {
        "properties": Object {
          "id": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "id",
        ],
        "type": "object",
      },
      "FooList": Object {
        "properties": Object {
          "count": Object {
            "nullable": false,
            "type": "integer",
          },
          "items": Object {
            "items": Object {
              "$ref": "#/components/schemas/Foo",
            },
            "nullable": false,
            "type": "array",
          },
          "limit": Object {
            "nullable": false,
            "type": "integer",
          },
          "offset": Object {
            "nullable": false,
            "type": "integer",
          },
        },
        "required": Array [
          "items",
          "count",
        ],
        "type": "object",
      },
      "QueryOutput": Object {
        "properties": Object {
          "response": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "response",
        ],
        "type": "object",
      },
      "ReplaceFoo": Object {
        "properties": Object {
          "bar": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "bar",
        ],
        "type": "object",
      },
      "UpdateFoo": Object {
        "properties": Object {
          "bar": Object {
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "bar",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/command": Object {
      "post": Object {
        "operationId": "mycommand",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/CommandInput",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/CommandOutput",
                },
              },
            },
            "description": "CommandOutput as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
      },
    },
    "/foo": Object {
      "get": Object {
        "operationId": "search",
        "parameters": Array [
          Object {
            "in": "query",
            "name": "offset",
            "schema": Object {
              "type": "integer",
            },
          },
          Object {
            "in": "query",
            "name": "limit",
            "schema": Object {
              "type": "integer",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/FooList",
                },
              },
            },
            "description": "FooList as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "head": Object {
        "operationId": "count",
        "parameters": Array [],
        "responses": Object {
          "204": Object {
            "description": "Empty response",
            "headers": Object {
              "X-Total-Count": Object {
                "schema": Object {
                  "type": "integer",
                },
              },
            },
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "post": Object {
        "operationId": "create",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/CreateFoo",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "201": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Foo",
                },
              },
            },
            "description": "Foo as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
    },
    "/foo/{fooId}": Object {
      "delete": Object {
        "operationId": "delete",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "Empty response",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "get": Object {
        "operationId": "retrieve",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Foo",
                },
              },
            },
            "description": "Foo as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "patch": Object {
        "operationId": "update",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/UpdateFoo",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Foo",
                },
              },
            },
            "description": "Foo as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
      "put": Object {
        "operationId": "replace",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/ReplaceFoo",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Foo",
                },
              },
            },
            "description": "Foo as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "foo",
        ],
      },
    },
    "/foo_file": Object {
      "post": Object {
        "operationId": "createfile",
        "parameters": Array [],
        "responses": Object {
          "201": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/FooFile",
                },
              },
            },
            "description": "FooFile as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "fooFile",
        ],
      },
    },
    "/foo_file/{fooFileId}": Object {
      "get": Object {
        "operationId": "retrievefile",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooFileId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "image/png": Object {
                "schema": Object {
                  "format": "binary",
                  "type": "string",
                },
              },
            },
            "description": "fooFile file content",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "fooFile",
        ],
      },
      "put": Object {
        "operationId": "replacefile",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "fooFileId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/FooFile",
                },
              },
            },
            "description": "FooFile as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "fooFile",
        ],
      },
    },
    "/query": Object {
      "get": Object {
        "operationId": "myquery",
        "parameters": Array [
          Object {
            "in": "query",
            "name": "request",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/QueryOutput",
                },
              },
            },
            "description": "QueryOutput as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
      },
    },
  },
  "servers": Array [
    Object {
      "url": "/",
    },
  ],
}
`;
